### PR TITLE
RFC2231 header parameter-value encoding

### PR DESCRIPTION
**Pull request by Maria-12648430**
As pointed out in #310, parameters are supposed to be encoded according to RFC2231, not RFC2047.

The value is first analyzed to decide which representation can/must be used, based on the presence of CTLs, spaces, tspecials, and UTF-8 multibyte characters. Then it is encoded in the selected representation. If the parameter lines become too long (>76 bytes), they are split up according to the continuation scheme described in RFC2231.

In hindsight, getting this right was surprisingly difficult, and therefore a careful review seems to be called for.

### CODE REVIEW

**Comment by seriyps** [comment]
Was surprized it did not trigger proper test failure, but it seems we don't generate long or non-ascii header parameter values.

https://github.com/gen-smtp/gen_smtp/blob/c6f25a758d60da9788bf5ddf73e985dacb28f74b/test/prop_mimemail.erl#L396

Might be nice to update when parser part is also implemented.

I'll have a deeper look bit later today, thak you!

**Comment by Maria-12648430** [comment]
Last commit takes care of an edge case: invalid UTF-8 sequences were not encoded but left as-is.
Also added specs.

**Comment by mworrell** [comment]
@seriyps Ok to merge this?

**Review by seriyps — APPROVED** [review]

**Comment by seriyps** [comment]
Great work! Sorry for the delay, it was due to UA-RU war.

**Comment by Maria-12648430** [comment]
@seriyps thanks 😄 TBH, I totally forgot about this one 😅

File: src/mimemail.erl [hunk 1]
```diff
@@ -916,28 +916,195 @@ has_lines_over_998(Bin, {FoundAt, 2}, _, Pattern) ->
...
+    C =:= $];
+    C =:= $?;
+    C =:= $=;
+    C =:= $\s;
+    C =:= $*;
+    C =:= $';
+    C =:= $%
+->
+    BinC = <<C/utf8>>,
+    EncSize = 3 * byte_size(BinC),
```
Comment (mworrell) [inline src/mimemail.erl:1044]: Just wondering. Given the allowed values in the when-clause, isn't the byte size of BinC always 1?

**Review by mworrell — APPROVED** [review]

File: src/mimemail.erl (hunk 1 above)
Comment (Maria-12648430) [reply src/mimemail.erl:1044]: Hm, you're right, and also the subsequent handling of the character is "off". I mean, it does the right thing, but overly complicated. I'm not quite sure what I was thinking back then I have to admit, given that I did that almost a year ago :sweat_smile: I probably restructured things at some point and forgot to pull it all straight. I'll look into it again, but probably not today. So don't merge this yet :wink:

Comment (Maria-12648430) [reply src/mimemail.erl]: Done :)

Comment (mworrell) [reply src/mimemail.erl]: That looks a lot simpler :-)

**Comment by mworrell** [comment]
@seriyps I am ok with this being merged.

**Pull request merged**